File: Practica1/Solution.go
package main

import (
	"fmt"
)

/*
 * Este método lo que hace es que dado un arreglo numérico multiplica cada
 * uno de sus elementos excepto por el í-esimo elemento, y el resultado lo
 * guardo en un arreglo de números enteros
 */

func productExceptSelf(nums []int) []int {
	/* Arreglo que se va a regresar con la respuesta*/
	answer := make([]int, len(nums))
	aux := 1
	aux2 := 1
	/*Slices para cuando recorremos de izquierda a derecha por primera vez el
	  arreglo y multiplicamos en dos en dos los elementos de este.*/
	firstrun := make([]int, 0)
	/*Segunda slice para cuando recorremos de derecha a izquierda por segunda
	  vez el arrgelo multiplicando en dos en dos los elementos de este*/
	secondrun := make([]int, 0)

	/*En este for recorremos de izquierda a derecha el arreglo dado en dos en dos
	  y el reultado se va guardando en la lista */
	for i := 0; i < len(nums); i++ {
		aux *= nums[i]
		firstrun = append(firstrun, aux)
	}

	/*En este for recorremos de derecha a izquierda el arreglo dado en dos en
	dos y el reultado se va guardando en la lista*/
	for i := len(nums) - 1; i >= 0; i-- {
		aux2 *= nums[i]
		secondrun = append(secondrun, aux2)
	}

	/*Volteamos la lista para poder multiplicar bien los elementos en sus posición*/
	reverse(secondrun)

	/*Recorremos el arreglo del parametro para poder identificar en que lugar
	estan los elementos y ver por cual número se debe de multiplicar el	i-esimo
	elemento menos 1 en la primera slice con el i-esimo elemento más 1 del segundo slice */
	for i := 0; i < len(nums); i++ {
		if i == 0 {
			answer[i] = secondrun[i+1]
		} else if i > 0 && i < len(nums)-1 {
			answer[i] = firstrun[i-1] * secondrun[i+1]
		} else {
			answer[i] = firstrun[i-1]
		}
	}

	return answer
}

/*Función que regresa la reversa de un arreglo*/
func reverse(nums []int) {
	for i := 0; i < len(nums)/2; i++ {
		nums[i], nums[len(nums)-i-1] = nums[len(nums)-i-1], nums[i]
	}
}

func main() {
	nums := []int{1, 2, 3, 4}
	fmt.Println(productExceptSelf(nums))
	nums2 := []int{-1, 1, 0, -3, 3}
	fmt.Println(productExceptSelf(nums2))
	nums3 := []int{-5, -2, -1, 1, 2, 7, 6}
	fmt.Println(productExceptSelf(nums3))
	nums4 := []int{0, 0, 0, 1, 2, 7, 0, 6}
	fmt.Println(productExceptSelf(nums4))
}

File: Practica1/Solution.java
package Practica1;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
    /*
     * @param un arreglo de números enteros
     * @return un arreglo de números enteros
     * Este método lo que hace es que dado un arreglo numérico multiplica cada 
     * uno de sus elementos excepto por el í-esimo elemento, y el reultado lo 
     * guardo en un arreglo de números enteros
     */
    public static int[] productExceptSelf(int[] nums) {
        int i; //Conatdor de los for
        int aux = 1; //Variable auxiliar 
        int aux2=1; //Variable auxiliar 2
        
        /*Lista para cuando recorremos de izquierda a derecha por primera vez el 
        arreglo y multiplicamos en dos en dos los elementos de este.
        Se utilizan ya que podemos acceder al i-esimo elemento con el método get() */
        List<Integer> list = new ArrayList<Integer>(); 
        /*Segunda lista para cuando recorremos de derecha a izquierda por segunda 
        vez el arrgelo multiplicando en dos en dos los elementos de este*/
        List<Integer> list2 = new ArrayList<Integer>();

        /* Arreglo que se va a regresar con la respuesta*/
        int[] answer = new int[nums.length];
        /*En este for recorremos de izquierda a derecha el arreglo dado en dos en dos 
        y el reultado se va guardando en la lista */
        for(i = 0 ; i < nums.length ; i++){
            aux = aux * nums[i];
            list.add(aux);
        }

        /*En este for recorremos de derecha a izquierda el arreglo dado en dos en 
        dos y el reultado se va guardando en la lista*/
        for(i = nums.length -1 ; i >= 0; i--){
            aux2 = aux2 * nums[i];
            list2.add(aux2);
        }

        /*Volteamos la lista2 para poder multiplicar bien los elementos */
        Collections.reverse(list2);

        /*Recorremos el arreglo del parametro para poder identificar en que lugar 
        estan los elementos y ver por cual cantidad se debe de multiplicar el el 
        i-esimo elemento menos 1 en la fila con el i-esimo elemento más 1 de la 
        segunda lista  */
        for(i = 0; i < nums.length; i++){
            if(i == 0){
                answer[i]=(int) list2.get(i+1);
            }
            else if( i > 0  && i < nums.length -1){
                answer[i]=list.get(i-1) * list2.get(i+1);
            }
            else{
                answer[i]=(int)list.get(i-1);
            }
        }

        return answer;
    }

}